components/web_discovery/browser/payload_generator_unittest.cc

```diff
@@ -0,0 +1,281 @@
+  // testing::Test:
+  void SetUp() override {
+    server_config_ = std::make_unique<ServerConfig>();
+    server_config_->location = "us";
+
+    url_details_ = std::make_unique<PatternsURLDetails>();
+
+    url_details_->payload_rule_groups = std::vector<PayloadRuleGroup>(2);
+
+    auto& single_group = url_details_->payload_rule_groups[0];
+    single_group.key = ".single-element";
+    single_group.rule_type = PayloadRuleType::kSingle;
+    single_group.result_type = PayloadResultType::kSingle;
+    single_group.action = "single_action";
+
+    auto& query_group = url_details_->payload_rule_groups[1];
+    query_group.key = "query_group";
+    query_group.rule_type = PayloadRuleType::kQuery;
+    query_group.result_type = PayloadResultType::kClustered;
+    query_group.action = "query";
+    query_group.rules = std::vector<PayloadRule>(2);
+
+    auto& join_rule = query_group.rules[0];
+    join_rule.selector = "#results";
+    join_rule.key = "r";
+    join_rule.is_join = true;
+
+    auto& qurl_rule = query_group.rules[1];
+    qurl_rule.selector = "qurl";
+    qurl_rule.key = "qurl";
+    qurl_rule.is_join = false;
+  }
+
+ protected:
+  std::vector<base::Value::Dict> GenerateQueryPayloadsHelper(
+      std::unique_ptr<PageScrapeResult> scrape_result) {
+    return GenerateQueryPayloads(*server_config_.get(), url_details_.get(),
+                                 std::move(scrape_result));
+  }
+
+  std::unique_ptr<ServerConfig> server_config_;
+
+ private:
+  std::unique_ptr<PatternsURLDetails> url_details_;
+};
+
+TEST_F(WebDiscoveryPayloadGeneratorTest, GenerateQueryPayloads) {
+  GURL test_url("https://example.com/test");
+  auto scrape_result = std::make_unique<PageScrapeResult>(test_url, "test_id");
+
+  std::vector<base::Value::Dict> single_dicts;
+  base::Value::Dict single_dict1;
+  single_dict1.Set("ab", "value1");
+  single_dict1.Set("cd", "value2");
+  single_dicts.push_back(std::move(single_dict1));
+  base::Value::Dict single_dict2;
+  single_dict2.Set("ef", "value3");
+  single_dict2.Set("gh", "value4");
+  single_dicts.push_back(std::move(single_dict2));
+  scrape_result->fields[".single-element"] = std::move(single_dicts);
+
+  std::vector<base::Value::Dict> result_dicts1;
+  base::Value::Dict result_dict1, result_dict2, result_dict3, result_dict4,
+      result_dict5;
+  result_dict1.Set("njk", "joinvalue1");
+  result_dict2.Set("abc", "joinvalue2");
+  result_dict3.Set("njk", "joinvalue3");
+  result_dict4.Set("abc", "joinvalue4");
+  result_dicts1.push_back(std::move(result_dict1));
+  result_dicts1.push_back(std::move(result_dict2));
+  result_dicts1.push_back(std::move(result_dict3));
+  result_dicts1.push_back(std::move(result_dict4));
+  std::vector<base::Value::Dict> result_dicts2;
+  result_dict5.Set("qurl", "https://example.com/test1");
+  result_dicts2.push_back(std::move(result_dict5));
+  scrape_result->fields["#results"] = std::move(result_dicts1);
+  scrape_result->fields["qurl"] = std::move(result_dicts2);
```

Comment:
maybe the whole thing is just `base::test::ParseJson`

Reply:
i found that the Dicts are probably a little too small for json parsing. decided to go with the Dict builder pattern that you mentioned below